### ISSUE TITLE
Support for visualizing pipeline with nested parallel stages

### ISSUE THREAD

**Issue by jemerald** [jenkinsci/pipeline-graph-view-plugin#51]
### Feature Request

Sample pipeline

```groovy
def myStage(name) {
    stage(name) {
        echo "$name"
    }
}

myStage('pre-parallel')

parallel p1: {
    myStage('p1-0')
    parallel 'p1-1-1': {
        myStage('p1-1-1')
    },'p1-1-2': {
        myStage('p1-1-2')
    }
    myStage('p1-2')
}, p2: {
    myStage('p2')
}

myStage('post-parallel')
```

Current graph:
![image](https://user-images.githubusercontent.com/9168432/139361125-134335b2-866e-4298-96a3-3cd0a455d5a6.png)

Desired graph:
```
pre-parallel --- p1-0 --- p1-1-1 --- p1-2 --- post-parallel
              |        \- p1-1-2 -/        |
              \-      p2                  -/
```

**Comment by jemerald** [comment]
@KalleOlaviNiemitalo yes, scripted pipeline only

@solvingj good point about "first-class" capability. I wonder whether the current limitation is due to the underlying datastructure, or just the UI render. There's a change that it's the former as the blueocean plugin also renders similar graph for the nested parallel pipeline.

**Comment by KalleOlaviNiemitalo** [comment]
From <https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/15978cd172a8d34e2cd6bf8a6bfeaa5514d44dad/rfc/JENKINS-41334-parallel-stage-execution.md>, it looks like nested parallel stages were disallowed in declarative pipelines to let Blue Ocean display the stages more easily. Perhaps that restriction could be removed if this plugin generally replaces the Blue Ocean visualisation.

**Comment by henryborchers** [comment]
I think disallowing nested parallel stages might be the single most limiting part of Jenkins. Removing this restriction would GREATLY improve my pipelines and would make declarative pipelines unparalleled powerful (sorry about the pun). 

I currently have to write awkward script blocks in my declarative pipelines to get the nesting parallel stages. This works fine enough to get what I need. I just wish that Jenkins could have nested parallel stages as a first class feature instead of having to use a hack (special nested scripted blocks).

**Comment by KalleOlaviNiemitalo** [comment]
[JENKINS-62773](https://issues.jenkins.io/browse/JENKINS-62773) requests allowing nested `parallel` stages.

[JENKINS-51808](https://issues.jenkins.io/browse/JENKINS-51808) requests allowing a dependency graph between stages that are executed in parallel. That could be useful for some jobs here but I imagine it would be tricky to visualize.

**Comment by PhotoTeeborChoka** [comment]
:+1:

**Comment by mdealer** [comment]
Maybe it's simpler to abandon the table-like visualization altogether and use something like a plain graph. There are libraries for this out there (but they are on NodeJS). I think it doesn't matter whether it fits the agenda or the religion. There are many janky parts of Jenkins where things don't really fit. It should just work and parallel stages are powerful and lacking in visualization.

**Comment by mark2185** [comment]
Sadly, it isn't, at least not for my case. I have nested parallel blocks and it shows only a few of them, precisely which is determined by a coin toss.

In addition to that, I cannot see all the "task nodes", and all the visible ones can turn green (and the banner turns green), as if the job is done, but some of the nodes are either waiting, working, or have failed.

**Comment by mark2185** [comment]
@PhotoTeeborChoka if you know a workaround, please let me know, I have to use a combination of both BlueOcean and `pipeline-graph-view` for this, but I'm still looking at a Jenkins-Frankenstein's monster.

`pipeline-graph-view` expands the parallel block vertically, but everything else is horizontal so I have about 300 nodes horizontally, and every 70 or so there's a vertical block of 15 'task nodes' (I don't know the terminology), but for viewing logs of a specific one I have to click on it, and then plug into the matrix and turn both keys at the same time to find it in the bottomless pit of the dropdown logs for each stage. Please tell me there's a better way.

**Comment by timja** [comment]
One thing I noticed recently is this pipeline here:
https://github.com/jenkinsci/docker-inbound-agent/blob/24600786d9342f27c3beac7f55a5e422dd2d360b/Jenkinsfile

which has parallel and matrix nested (using a combination of declarative and scripted) displays very broken while the build is running but fine after completion:
https://github.com/jenkinsci/docker-inbound-agent/blob/24600786d9342f27c3beac7f55a5e422dd2d360b/Jenkinsfile

would be good to see if we can fix that

**Comment by mdealer** [comment]
I guess a potential starting point for re-implementing the visualization using a graph would be taking a look at the ECharts API plugin, which somehow integrates nodejs into itself to provide ECharts (if I am not mistaken). As long as I am using Jenkins, it is only a question of time until I delve into this.

**Comment by galindroasml** [comment]
looking forward to see this feature implemented

**Comment by timbrown5** [comment]
I've been considering this as part of #233 - unfortunately, it's not a complete fix - I think the graph represented better, but I don't think the front end supports nested parallel blocks.

I've been trying to think through my expectations for a nested graph. I think for:
```
node {
    parallel([
        "A": {
            stage("Build") {
                echo("Build A")
            }
            stage("Test") {
                parallel([
                    "A1" : {
                        stage("Config") {
                            echo("Configuring test A1")
                        }
                        stage("Test") {
                            echo("Running test A1")
                        }
                    },
                    "A2" : {
                        stage("Config") {
                            echo("Configuring test A2")
                        }
                        stage("Test") {
                            echo("Running test A2")
                        }
                    }
                ])
            }
        },
        "B": {
            stage("Build") {
                echo("Build B")
            }
            parallel([
                "B1" : {
                    stage("Config") {
                        echo("Configuring test B1")
                    }
                    stage("Test") {
                        echo("Running test B1")
                    }
                },
                "B2" : {
                    stage("Config") {
                        echo("Configuring test B2")
                    }
                    stage("Test") {
                        echo("Running test B2")
                    }
                }
            ])
        }
    ])
}
```

I would expect something like this (albeit with the branches ordered down the page, A1, A2, B1, B2):
![graphviz(1)](https://user-images.githubusercontent.com/4447764/223727425-06047ef0-c5f1-4453-8c46-8a27e79fa8e5.svg)


You can see the [raw dot graph here](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20%20%20rankdir%3D%22LR%22%3B%0A%20%20%20%20newrank%3D%22true%22%0A%20%20%20%20Start%20%5Bgroup%3D%22main%22%5D%3B%0A%20%20%20%20Start%3Ae%20-%3E%2010%3Aw%20%5Blabel%20%3D%20%22A%22%5D%0A%20%20%20%20Start%3Ae%20-%3E%2012%3Aw%20%5Blabel%20%3D%20%22B%22%5D%0A%20%20%20%20End%20%5Bgroup%3D%22main%22%5D%3B%0A%20%20%20%2031%3Ae%20-%3E%20End%3Aw%0A%20%20%20%2033%3Ae%20-%3E%20End%3Aw%0A%20%20%20%2041%3Ae%20-%3E%20End%3Aw%0A%20%20%20%2043%3Ae%20-%3E%20End%3Aw%0A%20%20%20%20subgraph%20cluster_0%20%7B%0A%20%20%20%20%20%20%20%20rankdir%3D%22TB%22%3B%0A%20%20%20%20%20%20%20%20graph%5Bstyle%3Ddotted%5D%0A%20%20%20%20%20%20%20%20label%20%3D%20%22A%22%3B%0A%20%20%20%20%20%20%20%2010%20%5Blabel%3D%22Build%20A%22%2C%20group%3D%22main%22%5D%0A%20%20%20%20%20%20%20%2010%3Ae%20-%3E%2030%3Aw%20%5Blabel%20%3D%20%22Branch%3A%20A1%22%5D%0A%20%20%20%20%20%20%20%2010%3Ae%20-%3E%2032%3Aw%20%5Blabel%20%3D%20%22Branch%3A%20A2%22%5D%0A%20%20%20%20%20%20%20%20subgraph%20cluster_00%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20label%20%3D%20%22Test%22%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20graph%5Bstyle%3Ddotted%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2030%20%5Blabel%3D%22Config%22%2C%20group%3D%22main%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2031%20%5Blabel%3D%22Test%22%2C%20group%3D%22main%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2030%3Ae%20-%3E%2031%3Aw%0A%20%20%20%20%20%20%20%20%20%20%20%2032%20%5Blabel%3D%22Config%22%2C%20group%3D%22branch-A-A21%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2033%20%5Blabel%3D%22Test%22%2C%20group%3D%22branch-A-A2%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2032%3Ae%20-%3E%2033%3Aw%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%0A%20%20%20%20subgraph%20cluster_1%20%7B%0A%20%20%20%20%20%20%20%20rankdir%3D%22TB%22%3B%0A%20%20%20%20%20%20%20%20graph%5Bstyle%3Ddotted%5D%3B%0A%20%20%20%20%20%20%20%20label%20%3D%20%22B%22%3B%0A%20%20%20%20%20%20%20%2012%20%5Blabel%3D%22Build%20B%22%5D%0A%20%20%20%20%20%20%20%2012%3Ae%20-%3E%2040%3Aw%20%5Blabel%20%3D%20%22Branch%3A%20B1%22%5D%0A%20%20%20%20%20%20%20%2012%3Ae%20-%3E%2042%3Aw%20%5Blabel%20%3D%20%22Branch%3A%20B2%22%5D%0A%20%20%20%20%20%20%20%20subgraph%20cluster_10%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20graph%5Bstyle%3Ddotted%5D%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20label%20%3D%20%22Test%22%3B%0A%20%20%20%20%20%20%20%20%20%20%20%2040%20%5Blabel%3D%22Config%22%2C%20group%3D%22branch-B-B1%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2041%20%5Blabel%3D%22Test%22%2C%20group%3D%22branch-B-B1%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2040%3Ae%20-%3E%2041%3Aw%0A%20%20%20%20%20%20%20%20%20%20%20%2042%20%5Blabel%3D%22Config%22%2C%20group%3D%22branch-B-B2%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2043%20%5Blabel%3D%22Test%22%2C%20group%3D%22branch-B-B2%22%5D%0A%20%20%20%20%20%20%20%20%20%20%20%2042%3Ae%20-%3E%2043%3Aw%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%7D)

**Comment by galindroasml** [comment]
Understood. Please, keep us posted about it and enjoy your holiday!!

**Comment by galindroasml** [comment]
hey @timbrown5, did you enjoy your holidays? I hope so :)

Just a kind reminder about this feature. Did you have time to dive into this?

**Comment by asghaier** [comment]
Hi @timbrown5 ,

First of all thank you for the tree layout fix.

But I think PR #285 will not include a fix for this issue anymore.

Do you have an alternative idea to fix this issue? (maybe in another PR)

Thank you again for your contributions.

**Comment by henryborchers** [comment]
@timbrown5 any chance you'll give this another try?

**Comment by waclawski** [comment]
Hi, first thanks for your effort. Is there a chance to give this fix another try? Currently there is a responsibility circle beteween Blue Ocean and Stage View Plugin where both sides state that we should look for a solution on the other one. It looks like this plugin is our only chance to have a proper visualisation of complex, dynamic pipelines.

**Comment by gene-pavlovsky** [comment]
AFAIK Blue Ocean is dead, it's on life support / bugfixes only mode. So yes, better visualization here would be good to have :)

**Comment by SilverKnight1** [comment]
Hmm, howdy @timbrown5 I been looking at your #285, PipelineGraphLayout code but honestly for me its like playing a game of sticks. Change one thing and everything else freaks out. I'd be happy if just all nested nodes are visualized at the moment.

One of the biggest problems i keep running into while fiddling with code is that TreeVisitor/GraphAdapter code seems to translate pipelines differently depending on how you code your pipeline. I was trying to test Pure Declarative Pipelines with simple parallel stages and Pure Scripted Pipelines with nested parallels and Mixed half declarative with scripted nested parallels. (which is my preference)

The node data seems be structured per style/method of coding is not consistent. Somtimes nodes are not registered or should i say its not paired with its parents and so the visualizer is not given the information to make it appear. If you continue the project, maybe you can solve the data stream being consistant for all possible styles? 

If we could get all the methods to translate into the the same or similar node structure. Ideally in the order the pipeline is structured all into a single array. it would make it much easier to break down the nodestages and children of children of children nesting are way into infinity..etc. Right now sometimes it is...sometimes its not in order.

**Comment by Danoloan10** [comment]
I'm trying to refactor the react code a bit in my fork in https://github.com/Danoloan10/pipeline-graph-view-plugin with the goal that it makes it easier to make a nested graph view.

**Comment by henryborchers** [comment]
@Danoloan10 Awesome, if this feature gets implemented, it will encourage more efficient pipelines.

**Comment by agirault** [comment]
Is this project still active? Anyone assigned to this specific issue? The fact that nested parallel stages are not rendered right even though things work in the backend is unfortunate for diagnosis and debugging purposes.

**Comment by SilverKnight1** [comment]
From my understanding, it's kinda an overhaul hobby project for willing developers. Granted the product as is seems to be functional enough for most average pipelines but complex ones are harder to handle. But as i said before since the declarative and scripted coding styles results in different instructions to the graph visualizer its hard to update the graph visualizer to work for both styles. If you a seasoned developer be my guess. lol 

I wanted to get this working also but ran into the problem.  Every change to the graph visualizer resulted in one side coding style becoming even more broken.

**Comment by mdealer** [comment]
Found out about this a while ago: https://docs.cloudbees.com/docs/release-notes/latest/plugins/cloudbees-pipeline-explorer-plugin/

Apparently, it is kind of a Jenkins fork, or what am I looking at? Most of it smells like Jenkins but it looks different and feels closed. I could be mistaken, maybe someone can shed some light on this. It says it utilizes the Jenkins engine, but the rest is confusing. Does it mean we can jump to CloudBees, keep all of our existing Jenkins and utilize their commercial support and the advanced debugging plugins? Cannot wrap my head around how this is still open source.

**Comment by lemeurherve** [comment]
> Found out about this a while ago: https://docs.cloudbees.com/docs/release-notes/latest/plugins/cloudbees-pipeline-explorer-plugin/
>
> Apparently, it is kind of a Jenkins fork, or what am I looking at?

This CloudBees project is a proprietary one created from scratch, unrelated to this plugin or BlueOcean.

> It says it utilizes the Jenkins engine

It uses it to retrieve the logs.

To use it you have to be a CloudBees customer indeed.

**Comment by Danoloan10** [comment]
It's been a while since I've been able to get into this but I'll share some ideas I found.

@SilverKnight1 is right in that this will be a difficult feature to implement. The graph UI is designed as columns of parallel stages. This assumption can be found in all of the React code, and I have not yet found its bedrock. Because of this the redesign must be radical and change the way some things are being displayed. For example, stages within stages are not properly displayed even when they are sequential. And steps outside any stage in a scripted pipeline are not represetned at all. We will need something like what @timbrown5 designed in his dot diagram.

I've tried to model the high-level concepts that we're managing in this UML domain diagram:
![model](https://github.com/jenkinsci/pipeline-graph-view-plugin/assets/8666137/838e6874-6f04-4818-9b92-3ae6c4d51f48)
[click here for the plantuml](https://www.planttext.com/api/plantuml/txt/bLNBRiCW4BplLt2ld2-8gZJjCMeZvGietZGer7OnwQNDlvSO4WmiTkEB9NPczi6ElUeDr-RqgG24udtFjh939Hi4wBWuyWzaDTg8vaP-8_i1PX-7V0ddDn8PtPcBZADUdWVMy4-7FHSO-7N2naYk1cv95hj-Ccv3Ku2jVIk5FiSvTDmBBJk3TTJkpXyIlGwLh5KhZgF3DtD0JKQkIl-SoosnBtSYjDQy4OVx9HrlWtm_2th5tfwLkfcE3Wdid6LZKE-vGBRXxwYmVfPQACpFNShsPFBGaqOESBqCjlojyXkAejIwfWpMLe_ZsoGu0oJTCe1itMtAQcgbPmdnTfQLyap15tUnWbj8CTjuSJWZieDb0oerLe13oECgJTsKngCzAqPl5eejc7QMmZs0z4C6zXwSq_GmihoH7U-dUGl67UOpYjHB8A-L39yjgwgTKtFKR0oGEjjcNIpo2w9XHNj0z269DKE14o1Yz0IQTXcCFaa4BlWgmr73BeB8QuOLDlNbz_q7)

It's a bit complex, let me break it down:
- About labels:
  + LabeledCircle is the Stage circle we all know and love
  + LabeledBranchout is the labels that are floating on top of branchout curves (you can see them on top of arrows in @timbrown5's diagram)
  + LabeledArea would be @timbrown5's dotted subgraphs
- Consecutive steps that are neither `parallel` nor `stage` are grouped into StepsBlocks
- A ScriptedPipeline is just a collection of StepsBlocks. Thus there can be steps outside stages, as well as directly inside a stage with other stages in it.
- Circles would represent only pure StepsBlocks with no `parallel` or `stage`
- All other blocks and stages would be represented as Areas.

I hope this helps when me or anyone starts writing the new UI code, because it will need to be mostly done anew.

**Comment by nre-ableton** [comment]
Just a thought: maybe instead of trying to visualize nested-parallel pipelines in all their glory and complexity (which seems to be not only technically difficult, but unclear what the preferred design would be), could this plugin show nested parallel steps with a special circle icon, that when clicked would open those parallel stages elsewhere (the current view/new window/popup window)?

**Comment by timbrown5** [comment]
I think this is doable, the issue seems to be that
1) the frontend loops through all children to build the graph.
2) if a stage has children, it recurses into them,
- except when it sees a parallel branch it doesn't recurse, but handles these later.
3) for parallel branches it does a separate loop which doesn't recurse, so it ignores nested parallel branches.

If this was updated to loop through all stages and add parallel branches as rows and not new columns (like it does for stages) it would be a start (adding the paths might need to be updated to support this).

**Comment by Danoloan10** [comment]
Has the issue really been closed? #285's OP says otherwise

**Comment by timja** [comment]
Over eager closing fix words, the description had doesn’t fix issue ID which meant it got closed as it has fix issue ID

**Comment by galindro** [comment]
@Danoloan10 , are you currently working on the new UI code in order to address this improvement? Does #285 helps you to move forward?

BTW, ty very much for your efforts. 
Also kudos to @timbrown5 and @timja for merge #285

**Comment by Danoloan10** [comment]
I have not progressed much since the last time I commented, I will take
a look at the new code and rebase my fork, and see how it helps.
Thanks!

**Comment by timbrown5** [comment]
It should help with thinks like getting the correct  pipeline status and
timing (and correct parents). It likely won't make a difference with nested
parallel stages - as the graph structure is the same.
Feel free to put up a draft PR if you would like some feedback.

**Comment by timbrown5** [comment]
I looks at this a little and it seems straightforward to do what I said above and get the nested stages included (but not in the correct place/with the correct connections).
Before:
![Screenshot from 2024-03-12 09-14-12](https://github.com/jenkinsci/pipeline-graph-view-plugin/assets/4447764/a66420b5-7082-48d1-895b-a8686df70a3d)

After:
![Screenshot from 2024-03-12 09-12-23](https://github.com/jenkinsci/pipeline-graph-view-plugin/assets/4447764/366877be-77c3-4cf2-b2a5-6fcbffeae9eb)


(note that the 'After' stage is placed incorrectly and none of the nested stages are placed correctly or have the correct connections to the rest of the graph).

I took a loot and updating the placement and connection code, but it already seems convoluted and this might make it worse. It feels like there should be classes in this code to encapsulate all the data required to calculate connections/positions together (and split the large functions) - e.g. calculating a map of connections are required while looping through the stages.

Pipeline I used for testing:
```
node {
    parallel([
        "A": {
            stage("Build") {
                echo("Build A")
            }
            stage("Test") {
                parallel([
                    "A1" : {
                       echo("Test A1")
                    },
                    "A2" : {
                       echo("Test A2")
                    }
                ])
            }
        },
        "B": {
            stage("Build") {
                echo("Build B")
            }
            stage("Test") {
                parallel([
                    "B1" : {
                       echo("Test B1")
                    },
                    "B2" : {
                       echo("Test B2")
                    }
                ])
            }
        }
    ])
}
```

Added #360

**Comment by markojeremic** [comment]
Hello @timbrown5 and everyone else who is contributing to this,

What is the current status of this issue? Is it still in the progress?

I am pretty sure that you would make life easier to a lot of us here, so I am really happy if you continued to push forward towards this.

As always, thanks in advance for your efforts! :)

**Comment by porterctrlz** [comment]
This would be super useful!

**Comment by henryborchers** [comment]
Nested stages parallel works (almost) perfectly in Pipeline Console and has for a while now. Does anybody know what is needed to move forward for this to work in Pipeline Overview as well?

**Comment by Danoloan10** [comment]
AFAIK, the main pain point is rewriting the react code that is rendering the branches. The current column layout is deeply hardcoded, and also new interactions like collapsing parallel nodes need to be implemented. Appart from lacking experience in react programming, I have not managed to take the time to do it properly.

**Comment by lhupfeldt** [comment]
This would indeed be very welcome.

**Comment by myonlylonely** [comment]
Is this merged？

**Comment by myonlylonely** [comment]
Oh, I thought this is a PR.
